### ISSUE TITLE
Handlebars from yui as view-engine

### ISSUE THREAD

**Issue by focuzz8** [YahooArchive/mojito#270]
Since you want to merge components of YAF into mojito it makes sense to use handlebars as default (?) rendering engine.

**Comment by drewfish** [comment]
Yes, we're thinking about doing exactly that.  (See pull request #179.)  It'll take a little bit for that to fully happen, since all of our docs and examples are currently using mu.

**Comment by focuzz8** [comment]
I have client/server versions based on yui and this @caridy server version. I can make a pull request if you need.

**Comment by mridgway** [comment]
@focuzz You said you have a client side version of the Handlebars view engine. Do you have a link to this implementation?

**Comment by focuzz8** [comment]
@mridgway yes, i've uploaded it on github here https://github.com/focuzz/mojito/tree/hb

**Comment by focuzz8** [comment]
@mridgway i didn't found how compile works (server-side version), so it's just mimicry of mu version

**Comment by caridy** [comment]
@focuzz @mridgway yes, that's the main reason I decided not to pull the client side version of the HB implementation. YUI already does part of the job, and most important, HB relies heavily on partials and precompiled views. And this is not supported yet in mojito. @drewfish asked me to remove the changes that I did in mojito server side to precompile those views in HB way.

The way I see it, we should use this opportunity (renaming views into templates, etc) to embrase HB as the standard engine for mojito, and drop the current (and unnecessary) processes to load them from the client side in a form of *.mu.html directly, and instead, leverage YUI to do precompilation at the server side before flushing them to the client or merging them with the rest of the JS files through Shaker. And the current users, using mustache templates will not get affected since we can keep the backward compatibility.

**Comment by focuzz8** [comment]
sounds very good but how much time will it take? you want to rewrite hb compilation mechanics, right?

**Comment by caridy** [comment]
To clarify, that change that I'm proposing will remove the current mu/hb client side implementation, in favor of a more YUI friendly implementation, and it will obviously improve performance a lot since rendering templates will be a synchronous process without string manipulation.

**Comment by focuzz8** [comment]
@caridy you want compile templates directly into js code, and use this code on client and server directly, to make markup, right?

**Comment by caridy** [comment]
@focuzz yes. In theory, the resource store (/cc @drewfish) should be able to take care of this during the booting time (Shaker can also leverage resource store for the matter). In my opinion, this will bring few benefits:
- partials are easy when u already have pre-processed templates per mojits.
- easy to deploy the compiled templates, to a device, to a browser, to the server (by shaker)
- templates can be treated as YUI modules, hence we can define them as dependencies, etc. Very useful if you want to optimize what you load and when you load it.
- these new compiled templates will be familiar for anyone who has been using HB with YUI.
- and finally, the obvious performance improvement.

**Comment by drewfish** [comment]
In the past, it was an architectural decision that Mojito would allow multiple view engines, and not just one (mu).  Unless there's a strong argument otherwise, I'd like to keep this.  Keeping this flexibility means that users aren't stuck with handlebars if something better comes out in the future -- the user can use whichever they want.  (It's pretty easy to wire a new view engine into Mojito.)

FWIW, Mojito already supports compiling view by using `mojito compile views`.  This is something that needs support from each view engine (hb, mu, etc).  I agree it would be nice if Mojito did something like this automatically (to reduce the number of HTTP requests).

**Comment by caridy** [comment]
@drewfish right! but the reality is probably different. None of those templates engines have a chance against handlebar or dust for the web. Compiled templates for the client side is a must in the majority of the cases. Now, to address that concern, there is nothing that prevents us the use of a non-compilable template engine. We can alway send them into the client side in a form of TEXT, and then exposing them in a form of a function that will run in async mode to do the rest of the job. Here is an example with both approaches on it:

```
YUI.add("views/foo/bar", function (Y, NAME) {

      var templateObject = {/* a json representation of the template for non-compilable template engines */};
      /* or */
      var templateFunction = function (data) {/* a js representation of the template for compilable template engines */};

      YUI.namespace("_mojito._cache.compiled.' + mojit + '.views").bar = function (data, callback) {
           // do whatever you want here, if you have a templateFunction call it.
           callback(templateFunction(data));

           // or just use the templateObject if that's what u have with your engine running on top of YUI as a wrapper.
           Y.ABC.run(templateObj, data, callback);
      };

});
```

In other words, rather that provide support for engines, we can provide support for generating YUI modules that represent compiled templates that can be used at the client and server side alike. This aligns better with the fact that we can to delegate heavy string processes to the build or booting process for performance reasons as well. And at the end, a template is just a YUI module that expose a function under a certain namespace, and that function expect data and a callback.

**Comment by drewfish** [comment]
I think I understand what you're saying.  In today's Mojito, you should be able to create a view engine that implements exactly what you describe :)

**Comment by mridgway** [comment]
This was released in 0.4.2.
